index.js

```diff
@@ -45,115 +82,192 @@ const propertiesISOEquivalent = {
-  return toISOString.call(this, { short: false })
-}
+PostgresInterval.prototype.toISOString = PostgresInterval.prototype.toISO =
+  function () {
+    return toISOString.call(this, { short: false })
+  }
 
 PostgresInterval.prototype.toISOStringShort = function () {
   return toISOString.call(this, { short: true })
 }
 
-function toISOString ({ short = false }) {
-  const datePart = dateProperties
-    .map(buildProperty, this)
-    .join('')
+function toISOString ({ short }) {
+  let datePart = ''
+
+  if (!short || this.years) {
+    datePart += this.years + propertiesISOEquivalent.years
+  }
+
+  if (!short || this.months) {
+    datePart += this.months + propertiesISOEquivalent.months
+  }
+
+  if (!short || this.days) {
+    datePart += this.days + propertiesISOEquivalent.days
+  }
+
+  let timePart = ''
+
+  if (!short || this.hours) {
+    timePart += this.hours + propertiesISOEquivalent.hours
+  }
+
+  if (!short || this.minutes) {
+    timePart += this.minutes + propertiesISOEquivalent.minutes
+  }
 
-  const timePart = timeProperties
-    .map(buildProperty, this)
-    .join('')
+  if (!short || (this.seconds || this.milliseconds)) {
+    if (this.milliseconds) {
+      timePart += (Math.trunc((this.seconds + this.milliseconds / 1000) * 1000000) / 1000000) + propertiesISOEquivalent.seconds
+    } else {
+      timePart += this.seconds + propertiesISOEquivalent.seconds
+    }
+  }
 
-  if (!timePart.length && !datePart.length) return 'PT0S'
+  if (!timePart && !datePart) {
+    return 'PT0S'
+  }
 
-  if (!timePart.length) return `P${datePart}`
+  if (!timePart) {
+    return `P${datePart}`
+  }
 
   return `P${datePart}T${timePart}`
+}
+
+const position = { value: 0 }
+
+function readNextNum (interval) {
+  let val = 0
 
-  function buildProperty (property) {
-    let value = this[property] || 0
+  while (position.value < interval.length) {
+    const char = interval[position.value]
 
-    // Account for fractional part of seconds,
-    // remove trailing zeroes.
-    if (property === 'seconds' && this.milliseconds) {
-      value = (value + this.milliseconds / 1000).toFixed(6).replace(/0+$/, '')
+    if (char >= '0' && char <= '9') {
+      val = val * 10 + +char
+      position.value++
+    } else {
+      break
     }
+  }
+
+  return val
+}
+
+function parseMillisecond (interval) {
+  const currentValue = readNextNum(interval)
+
+  if (currentValue < 10) {
+    return currentValue * 100
+  }
+
+  if (currentValue < 100) {
+    return currentValue * 10
+  }
+
+  if (currentValue < 1000) {
+    return currentValue
+  }
+
+  if (currentValue < 10000) {
+    return currentValue / 10
+  }
 
-    if (short && !value) return ''
+  if (currentValue < 100000) {
+    return currentValue / 100
+  }
 
-    return value + propertiesISOEquivalent[property]
+  if (currentValue < 1000000) {
+    return currentValue / 1000
   }
+
+  // slow path
+  const remainder = currentValue.toString().length - 3
+  return currentValue / remainder
```

Comment:
There’s no way this is right, right? Needs test coverage.

Reply:
Good catch, is missing `currentValue / Math.pow(10, remainder)`,

Actually, it is impossible for the code to reach this condition because milliseconds will not pass more than 6 digits, but I will fix it.

Reply:
Bug fixed and tests fixed!